File: java/gadgets/src/main/java/org/apache/shindig/gadgets/RemoteContentRequest.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.shindig.gadgets;

import java.io.ByteArrayInputStream;
import java.io.InputStream;
import java.net.URI;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Holds request data for passing to a RemoteContentFetcher.
 * Instances of this object are immutable.
 *
 * TODO: We should probably just stick the method in here. Having separate
 * calls for POST vs. get seems unnecessary.
 *
 * TODO: This naming seems really ridiculous now. Why don't we just call it
 * what it is -- an HTTP request?
 */
public class RemoteContentRequest {
  private final byte[] postBody;

  /**
   * @return An input stream that can be used to read the post body.
   */
  public InputStream getPostBody() {
    return new ByteArrayInputStream(postBody);
  }

  /**
   * Retrieves the total length of the post body.
   *
   * @return The length of the post body.
   */
  public int getPostBodyLength() {
    return postBody.length;
  }

  private final String contentType;
  private final static String DEFAULT_CONTENT_TYPE
      = "application/x-www-form-urlencoded; charset=utf-8";

  /**
   * @return The content type of the request (determined from request headers)
   */
  public String getContentType() {
    return contentType;
  }

  private final Map<String, List<String>> headers;

  /**
   * @return All headers set in this request.
   */
  public Map<String, List<String>> getAllHeaders() {
    return headers;
  }

  /**
   * @param name The header to fetch
   * @return A list of headers with that name (may be empty).
   */
  public List<String> getHeaders(String name) {
    List<String> match = headers.get(name);
    if (match == null) {
      return Collections.emptyList();
    } else {
      return match;
    }
  }

  /**
   * @param name
   * @return The first set header with the given name or null if not set. If
   *         you need multiple values for the header, use getHeaders().
   */
  public String getHeader(String name) {
    List<String> headerList = getHeaders(name);
    if (headerList.size() == 0) {
      return null;
    } else {
      return headerList.get(0);
    }
  }

  private final String method;
  public String getMethod() {
    return method;
  }

  private final URI uri;
  public URI getUri() {
    return uri;
  }

  private final Options options;
  public Options getOptions() {
    return options;
  }

  /**
   *
   * @param method
   * @param uri
   * @param headers
   * @param postBody
   * @param options
   */
  public RemoteContentRequest(String method,
                              URI uri,
                              Map<String, List<String>> headers,
                              byte[] postBody,
                              Options options) {
    this.method = method;
    this.uri = uri;
    this.options = options;
    // Copy the headers
    if (headers == null) {
      this.headers = Collections.emptyMap();
    } else {
      Map<String, List<String>> tmpHeaders
          = new HashMap<String, List<String>>();
      for (Map.Entry<String, List<String>> entry : headers.entrySet()) {
        List<String> newList = new ArrayList<String>(entry.getValue());
        tmpHeaders.put(entry.getKey(), Collections.unmodifiableList(newList));
      }
      this.headers = Collections.unmodifiableMap(tmpHeaders);
    }
    if (postBody == null) {
      this.postBody = new byte[0];
    } else {
      this.postBody = new byte[postBody.length];
      System.arraycopy(postBody, 0, this.postBody, 0, postBody.length);
    }

    // Calculate content type.
    String type = getHeader("Content-Type");
    if (type == null) {
      contentType = DEFAULT_CONTENT_TYPE;
    } else {
      contentType = type;
    }
  }

  /**
   * Basic GET request.
   *
   * @param uri
   */
  public RemoteContentRequest(URI uri) {
    this("GET", uri, null, null, DEFAULT_OPTIONS);
  }

  /**
   * GET with options
   *
   * @param uri
   * @param options
   */
  public RemoteContentRequest(URI uri, Options options) {
    this("GET", uri, null, null, options);
  }

  /**
   * GET request with custom headers and default options
   * @param uri
   * @param headers
   */
  public RemoteContentRequest(URI uri, Map<String, List<String>> headers) {
    this("GET", uri, headers, null, DEFAULT_OPTIONS);
  }

  /**
   * GET request with custom headers + options
   * @param uri
   * @param headers
   * @param options
   */
  public RemoteContentRequest(URI uri, Map<String, List<String>> headers,
      Options options) {
    this("GET", uri, headers, null, options);
  }

  /**
   * Basic POST request
   * @param uri
   * @param postBody
   */
  public RemoteContentRequest(URI uri, byte[] postBody) {
    this("POST", uri, null, postBody, DEFAULT_OPTIONS);
  }

  /**
   * POST request with options
   * @param uri
   * @param postBody
   * @param options
   */
  public RemoteContentRequest(URI uri, byte[] postBody, Options options) {
    this("POST", uri, null, postBody, options);
  }

  /**
   * POST request with headers
   * @param uri
   * @param headers
   * @param postBody
   */
  public RemoteContentRequest(URI uri, Map<String, List<String>> headers,
      byte[] postBody) {
    this("POST", uri, headers, postBody, DEFAULT_OPTIONS);
  }

  /**
   * POST request with options + headers
   * @param uri
   * @param headers
   * @param postBody
   * @param options
   */
  public RemoteContentRequest(URI uri, Map<String, List<String>> headers,
      byte[] postBody, Options options) {
    this("POST", uri, headers, postBody, options);
  }

  @Override
  public String toString() {
    StringBuilder buf = new StringBuilder();
    buf.append(method).append(" ").append(uri.getPath()).append(" HTTP/1.1\r\n")
       .append("Host: ").append(uri.getHost())
       .append(uri.getPort() == 80 ? "" : ":" + uri.getPort())
       .append("\r\n");

    for (Map.Entry<String, List<String>> entry : headers.entrySet()) {
      buf.append(entry.getKey()).append(": ");
      boolean first = false;
      for (String header : entry.getValue()) {
        if (!first) {
          first = true;
        } else {
          buf.append(", ");
        }
        buf.append(header);
      }
      buf.append("\r\n");
    }
   buf.append("\r\n");
   buf.append(new String(postBody));
   return buf.toString();
  }

  @Override
  public boolean equals(Object rhs) {
    if (rhs == this) {return true;}
    if (rhs instanceof RemoteContentRequest) {
      RemoteContentRequest req = (RemoteContentRequest)rhs;
      return method.equals(req.method) &&
             uri.equals(req.uri) &&
             Arrays.equals(postBody, req.postBody) &&
             headers.equals(req.headers);
    }
    return false;
  }

  public static final Options DEFAULT_OPTIONS = new Options();

  /**
   * Bag of options for making a request.
   *
   * This object is mutable to keep us sane. Don't mess with it once you've
   * sent it to RemoteContentRequest or bad things might happen.
   */
  public static class Options {
    public boolean ignoreCache = false;
  }
}

File: 1.0.x/php/src/common/RemoteContentRequest.php
<?php
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */


class RemoteContentRequest {
  // these are used for making the request
  private $uri = '';
  // to get real url after signed requests
  private $notSignedUri = '';
  private $method = '';
  private $headers = array();
  private $postBody = false;
  // these fields are filled in once the request has completed
  private $responseContent = false;
  private $responseSize = false;
  private $responseHeaders = false;
  private $httpCode = false;
  private $contentType = null;
  private $options;
  private $created;
  private static $SC_OK = 200; //Please, use only for testing!
  public $handle = false;
  public static $DEFAULT_CONTENT_TYPE = "application/x-www-form-urlencoded; charset=utf-8";

  public function __construct($uri, $headers = false, $postBody = false) {
    $this->uri = $uri;
    $this->notSignedUri = $uri;
    $this->headers = $headers;
    $this->postBody = $postBody;
    $this->created = time();
  }

  public function createRemoteContentRequest($method, $uri, $headers, $postBody, $options) {
    $this->method = $method;
    $this->uri = $uri;
    $this->options = $options;
    // Copy the headers
    if (! isset($headers)) {
      $this->headers = '';
    } else {
      $setPragmaHeader = false;
      $tmpHeaders = '';
      foreach ($headers as $key => $value) {
        // Proxies should be bypassed with the Pragma: no-cache check.
        if ($key == "Pragma" && $options->ignoreCache) {
          $value = "no-cache";
          $setPragmaHeader = true;
        }
        $tmpHeaders .= $key . ":" . $value . "\n";
      }
      // Bypass caching in proxies as well.
      if (! $setPragmaHeader && $options->ignoreCache) {
        $tmpHeaders .= "Pragma:no-cache\n";
      }
      $this->headers = $tmpHeaders;
    }
    if (! isset($postBody)) {
      $this->postBody = '';
    } else {
      $this->postBody = array_merge($postBody, $this->postBody);
    }
    $type = $this->getHeader("Content-Type");
    if (! isset($type)) {
      $this->contentType = RemoteContentRequest::$DEFAULT_CONTENT_TYPE;
    } else {
      $this->contentType = $type;
    }
  }

  /**
   * Creates a new request to a different URL using all request data from
   * an existing request.
   *
   * @param uri
   * @param base The base request to copy data from.
   */
  public static function createRemoteContentRequestWithUriBase($uri, $base) {
    $this->uri = $uri;
    $this->method = $base->method;
    $this->options = $base->options;
    $this->headers = $base->headers;
    $this->contentType = $base->contentType;
    $this->postBody = $base->postBody;
  }

  /**
   * Basic GET request.
   *
   * @param uri
   */
  public function createRemoteContentRequestWithUri($uri) {
    $this->createRemoteContentRequest("GET", $uri, null, null, RemoteContentRequest::getDefaultOptions());
  }

  /**
   * GET with options
   *
   * @param uri
   * @param options
   */
  public function createRemoteContentRequestWithUriOptions($uri, $options) {
    $this->createRemoteContentRequest("GET", $uri, null, null, $options);
  }

  /**
   * GET request with custom headers and default options
   * @param uri
   * @param headers
   */
  public function RemoteContentRequestWithUriHeaders($uri, $headers) {
    $this->createRemoteContentRequest("GET", $uri, $headers, null, RemoteContentRequest::getDefaultOptions());
  }

  /**
   * GET request with custom headers + options
   * @param uri
   * @param headers
   * @param options
   */
  public function createRemoteContentRequestWithUriHeadersOptions($uri, $headers, $options) {
    $this->createRemoteContentRequest("GET", $uri, $headers, null, $options);
  }

  /**
   * Basic POST request
   * @param uri
   * @param postBody
   */
  public function RemoteContentRequestWithUriPostBody($uri, $postBody) {
    $this->createRemoteContentRequest("POST", $uri, null, $postBody, RemoteContentRequest::getDefaultOptions());
  }

  /**
   * POST request with options
   * @param uri
   * @param postBody
   * @param options
   */
  public function createRemoteContentRequestWithUriPostBodyOptions($uri, $postBody, $options) {
    $this->createRemoteContentRequest("POST", $uri, null, $postBody, $options);
  }

  /**
   * POST request with headers
   * @param uri
   * @param headers
   * @param postBody
   */
  public function createRemoteContentRequestWithUriHeadersPostBody($uri, $headers, $postBody) {
    $this->createRemoteContentRequest("POST", $uri, $headers, $postBody, RemoteContentRequest::getDefaultOptions());
  }

  /**
   * POST request with options + headers
   * @param uri
   * @param headers
   * @param postBody
   * @param options
   */
  public function createRemoteContentRequestWithUriHeadersPostBodyOptions($uri, $headers, $postBody, $options) {
    $this->createRemoteContentRequest("POST", $uri, $headers, $postBody, $options);
  }

  /**
   * Creates a simple GET request
   *
   * @param uri
   * @param ignoreCache
   */
  public function getRequest($uri, $ignoreCache) {
    $options = new Options();
    $options->ignoreCache = $ignoreCache;
    return $this->createRemoteContentRequestWithUriOptions($uri, $options);
  }

  /**
   * Simple constructor for setting a basic response from a string. Mostly used
   * for testing.
   *
   * @param body
   */
  public function getHttpFalseResponseBody($body) {
    return $this->createFalseResponse(RemoteContentRequest::$SC_OK, $body, null);
  }

  private function createFalseResponse($httpCode, $body, $headers) {
    $this->httpCode = $httpCode;
    $this->responseContent = $body;
    $this->headers = $headers;
    return $this;
  }

  // returns a hash code which identifies this request, used for caching
  // takes url and postbody into account for constructing the sha1 checksum
  public function toHash() {
    return md5($this->uri . $this->postBody);
  }

  public static function getDefaultOptions() {
    return new Options();
  }

  public function getContentType() {
    return $this->contentType;
  }

  public function getHttpCode() {
    return $this->httpCode;
  }

  public function getResponseContent() {
    return $this->responseContent;
  }

  public function getResponseHeaders() {
    return $this->responseHeaders;
  }

  public function getResponseSize() {
    return $this->responseSize;
  }

  public function getHeaders() {
    return $this->headers;
  }

  public function isPost() {
    return ($this->postBody != false);
  }

  public function hasHeaders() {
    return ! empty($this->headers);
  }

  public function getPostBody() {
    return $this->postBody;
  }

  public function getUrl() {
    return $this->uri;
  }

  public function getNotSignedUrl() {
    return $this->notSignedUri;
  }

  public function getMethod() {
    return $this->method;
  }

  public function getOptions() {
    if (empty($this->options)) {
      return new Options();
    }
    return $this->options;
  }

  public function setContentType($type) {
    $this->contentType = $type;
  }

  public function setHttpCode($code) {
    $this->httpCode = intval($code);
  }

  public function setResponseContent($content) {
    $this->responseContent = $content;
  }

  public function setResponseHeaders($headers) {
    $this->responseHeaders = $headers;
  }

  public function setResponseSize($size) {
    $this->responseSize = intval($size);
  }

  public function setHeaders($headers) {
    $this->headers = $headers;
  }

  //FIXME: Find a better way to do this
  // The headers can be an array of elements.
  public function getHeader($headerName) {
    $headers = explode("\n", $this->headers);
    foreach ($headers as $header) {
      $key = explode(":", $header, 2);
      if ($key[0] == $headerName) return trim($key[1]);
    }
    return null;
  }

  //FIXME: Find a better way to do this
  // The headers can be an array of elements.
  public function getResponseHeader($headerName) {
    $headers = explode("\n", $this->responseHeaders);
    foreach ($headers as $header) {
      $key = explode(":", $header, 2);
      if ($key[0] == $headerName) {
        return trim($key[1]);
      }
    }
    return null;
  }

  public function getCreated() {
    return $this->created;
  }

  public function setPostBody($postBody) {
    $this->postBody = $postBody;
  }

  public function setUri($uri) {
    $this->uri = $uri;
  }

  public function setNotSignedUri($uri) {
    $this->notSignedUri = $uri;
  }
}

/**
 * Bag of options for making a request.
 *
 * This object is mutable to keep us sane. Don't mess with it once you've
 * sent it to RemoteContentRequest or bad things might happen.
 */
class Options {
  public $ignoreCache = false;
  public $ownerSigned = true;
  public $viewerSigned = true;

  public function __construct() {}

  /**
   * Copy constructor
   */
  public function copyOptions(Options $copyFrom) {
    $this->ignoreCache = $copyFrom->ignoreCache;
    $this->ownerSigned = $copyFrom->ownerSigned;
    $this->viewerSigned = $copyFrom->viewerSigned;
  }

}
